WalletWasabi/Services/WalletService.cs

```diff
@@ -465,27 +451,51 @@ public List<SmartCoin> GetClusters(SmartCoin coin, List<SmartCoin> current, ILoo
 			return clusters;
 		}
 
-		private void ProcessBlock(Height height, Block block)
+		private void ProcessBlock(Height height, Block block, IEnumerable<int> filterByTxIds = null)
 		{
-			foreach (var tx in block.Transactions)
+			if (filterByTxIds is null)
+			{
+				var relevantIndicies = new List<int>();
+				for (int i = 0; i < block.Transactions.Count; i++)
+				{
+					Transaction tx = (Transaction)block.Transactions[i];
+					if (ProcessTransaction(new SmartTransaction(tx, height)))
+					{
+						relevantIndicies.Add(i);
+					}
+				}
+
+				if (relevantIndicies.Any())
+				{
+					var blockState = new BlockState(block.GetHash(), height, relevantIndicies);
+					KeyManager.AddBlockState(blockState);
+				}
+				KeyManager.SetBestHeight(height);
+			}
+			else
 			{
-				ProcessTransaction(new SmartTransaction(tx, height));
+				foreach (var i in filterByTxIds.OrderBy(x => x))
```

Comment:
I think the transaction indexes are already ordered here. Also, I would change the parameter name to `filterByTransactionIndexes` or something like that because `TxIds` is used for transaction identifiers (uint256).

```suggestion
				foreach (var i in filterByTxIds)
```

Reply:
You're right, I've been thinking on this one, but I'm afraid to remove this. It's crucial that `ProcessBlock` to go through txs in order.

Reply:
I'd leave it here. It's not a performance hit and I can sleep better at night with this:)